binary-transcoders/src/main/kotlin/com/kamelia/sprinkler/codec/binary/encoder/EncoderUtils.kt

```diff
@@ -0,0 +1,181 @@
+@file:JvmName("EncoderUtils")
+
+package com.kamelia.sprinkler.codec.binary.encoder
+
+import com.kamelia.sprinkler.codec.binary.encoder.core.Encoder
+
+/**
+ * Creates an encoder that encodes an object [R] using a given encoder of type [T]. The created encoder will map the
+ * input object [R] to an object of type [T] using the given [mapper] function, and then delegate the encoding to the
+ * original encoder.
+ *
+ * Here is an example of how to use this function:
+ *
+ * ```
+ * val longEncoder = LongEncoder()
+ * val instantEncoder: Encoder<Instant> = longEncoder.withMappedInput(Instant::toEpochMilli)
+ * ```
+ *
+ * The above code creates an encoder that encodes [Instant][java.time.Instant] objects by first converting them to a
+ * long value using [Instant.toEpochMilli][java.time.Instant.toEpochMilli], and then encoding the long value using the
+ * longEncoder.
+ *
+ * @receiver the original encoder of type [T]
+ * @param mapper the function that maps the input object [R] to an object of type [T]
+ * @return an encoder that encodes an object [R]
+ */
+fun <T, R> Encoder<T>.withMappedInput(mapper: (R) -> T): Encoder<R> =
+    Encoder { obj, output -> this@withMappedInput.encode(mapper(obj), output) }
```

Comment:
```suggestion
    Encoder { obj, output -> encode(mapper(obj), output) }
```

same for almost all the this@tag in this file